README.md
# Flutter Twitch Player

![](https://variety.com/wp-content/uploads/2015/03/twitch-logo.jpg?w=636)

A flutter plugin that plays twitch live streams using [InAppWebview](https://pub.dev/packages/flutter_inappwebview) under the hood.

## Getting started

Update gradle in `android/build.gradle`

```gradle
dependencies {
classpath 'com.android.tools.build:gradle:7.0.0'
}
```

Update minSDK to 17 in `android/app/build.gradle`

```gradle
defaultConfig {
    minSdkVersion 17
}
```

Update distributionUrl in `android/gradle/gradle-wrapper.properties`

```gradle
distributionUrl=https\://services.gradle.org/distributions/gradle-7.0.2-all.zip
```

## Usage

```dart
TwitchPlayerIFrame(channel: "channel name", controller: TwitchController())
```

## TODO

- play
- pause
- mute
- unmute
- fullscreen

## Additional information

Only tested on android.


CHANGELOG.md
## 0.1.4

- Added autoplay param
- Added a controller class
- Update README

## 0.0.4

- Added more comments
- Update Changelog
- Update README
- Initial Commit


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_twitch_player/flutter_twitch_player.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Twitch Player Example',
      theme: ThemeData(
        primarySwatch: Colors.purple,
      ),
      home: const MyHomePage(title: 'Twitch Player Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  TwitchController twitchController = TwitchController();
  @override
  Widget build(BuildContext context) {
    twitchController.onEnterFullscreen(() => print("Entered fullscreen"));
    twitchController.onExitFullscreen(() => print("Exited fullscreen"));
    twitchController.onStateChanged((state) => print(state));
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
        centerTitle: true,
      ),
      body: Column(
        children: [
          Padding(
            padding: const EdgeInsets.all(16),
            child: TwitchPlayerIFrame(
              controller: twitchController,
              channel: "Crown",
              borderRadius: BorderRadius.circular(6),
            ),
          ),
        ],
      ),
    );
  }
}


lib/src/flutter_twitch_player.dart
import 'package:flutter/material.dart';
import 'package:flutter_inappwebview/flutter_inappwebview.dart';

import 'flutter_twitch_controller.dart';

class TwitchPlayerIFrame extends StatefulWidget {
  /// Inline Twitch Video Player
  const TwitchPlayerIFrame({
    Key? key,
    required this.channel,
    required this.controller,
    this.parent = "twitch.tv",
    this.autoplay = true,
    this.borderRadius,
  }) : super(key: key);

  /// Twitch channel name
  final String channel;

  /// Controller
  final TwitchController controller;

  /// Twitch parent, usually twitch.tv
  final String parent;

  /// Autoplay the video
  final bool autoplay;

  /// Border radius if desired
  final BorderRadius? borderRadius;

  @override
  State<TwitchPlayerIFrame> createState() => _TwitchPlayerIFrameState();
}

class _TwitchPlayerIFrameState extends State<TwitchPlayerIFrame> {
  late InAppWebViewController inAppWebViewController;

  @override
  Widget build(BuildContext context) {
    return AspectRatio(
      aspectRatio: 16 / 9,
      child: ClipRRect(
        borderRadius: widget.borderRadius ?? BorderRadius.zero,
        child: InAppWebView(
          initialUrlRequest: URLRequest(
            url: Uri(
              scheme: 'https',
              host: 'player.twitch.tv',
              queryParameters: {
                "channel": widget.channel,
                "parent": widget.parent,
              },
            ),
          ),
          onEnterFullscreen: (_) => widget.controller.enterFullscreen(),
          onExitFullscreen: (_) => widget.controller.exitFullscreen(),
          onLoadStart: (controller, __) => {
            inAppWebViewController = controller,
            inAppWebViewController.addUserScript(
              userScript: UserScript(source: """
                async function wait() {
                  while(!document.getElementsByClassName('ScLoadingSpinnerCircle-sc-qd487d-1')[0]) {
                    await new Promise(r => setTimeout(r, 500));
                  }
                  while(document.getElementsByClassName('ScLoadingSpinnerCircle-sc-qd487d-1')[0]) {
                    await new Promise(r => setTimeout(r, 500));
                  }
                  ${widget.autoplay ? """""" : """document.querySelector('[data-a-target="player-play-pause-button"]').click();"""}
                }
                wait();
              """, injectionTime: UserScriptInjectionTime.AT_DOCUMENT_END),
            ),
            widget.controller.updateState(TwitchState.loading),
          },
          onLoadStop: (_, __) => {
            widget.controller.updateState(TwitchState.done),
          },
          onLoadError: (_, __, ___, ____) =>
              widget.controller.updateState(TwitchState.error),
        ),
      ),
    );
  }
}


lib/src/flutter_twitch_controller.dart
class TwitchController {
  /// Twitch Video Player Controller
  bool isFullscreen = false;
  String twitchState = TwitchState.loading;
  Function? _onEnterFullscreenFunction;
  Function? _onExitFullscreenFunction;
  late void Function(String) _onStateChanged;

  void onEnterFullscreen(Function function) {
    _onEnterFullscreenFunction = function;
  }

  void onExitFullscreen(Function function) {
    _onExitFullscreenFunction = function;
  }

  void onStateChanged(void Function(String state) function) {
    _onStateChanged = function;
  }

  void enterFullscreen() async {
    isFullscreen = true;
    if (_onEnterFullscreenFunction != null) {
      _onEnterFullscreenFunction?.call();
    }
  }

  void exitFullscreen() async {
    isFullscreen = false;
    if (_onExitFullscreenFunction != null) {
      _onExitFullscreenFunction?.call();
    }
  }

  void updateState(String state) {
    twitchState = state;
    _onStateChanged.call(state);
  }
}

class TwitchState {
  static const loading = "LOADING";
  static const done = "DONE";
  static const error = "ERROR";
}


lib/src/flutter_twitch_iframe.dart
import 'package:flutter/material.dart';
import 'package:flutter_inappwebview/flutter_inappwebview.dart';

import 'flutter_twitch_controller.dart';

class TwitchPlayerIFrame extends StatefulWidget {
  /// Inline Twitch Video Player
  const TwitchPlayerIFrame({
    Key? key,
    required this.channel,
    required this.controller,
    this.parent = "twitch.tv",
    this.autoplay = true,
    this.borderRadius,
  }) : super(key: key);

  /// Twitch channel name
  final String channel;

  /// Controller
  final TwitchController controller;

  /// Twitch parent, usually twitch.tv
  final String parent;

  /// Starting volume
  // final double volume;

  // /// Start muted?
  // final bool muted;

  /// Autoplay the video
  final bool autoplay;

  /// Autoplay the video
  // final bool fullscreen;

  /// Border radius if desired
  final BorderRadius? borderRadius;

  @override
  State<TwitchPlayerIFrame> createState() => _TwitchPlayerIFrameState();
}

class _TwitchPlayerIFrameState extends State<TwitchPlayerIFrame> {
  late InAppWebViewController inAppWebViewController;

  @override
  Widget build(BuildContext context) {
    return AspectRatio(
      aspectRatio: 16 / 9,
      child: ClipRRect(
        borderRadius: widget.borderRadius ?? BorderRadius.zero,
        child: InAppWebView(
          initialUrlRequest: URLRequest(
            url: Uri(
              scheme: 'https',
              host: 'player.twitch.tv',
              queryParameters: {
                "channel": widget.channel,
                "parent": widget.parent,
              },
            ),
          ),
          onEnterFullscreen: (_) => widget.controller.enterFullscreen(),
          onExitFullscreen: (_) => widget.controller.exitFullscreen(),
          onLoadStart: (controller, __) => {
            inAppWebViewController = controller,
            inAppWebViewController.addUserScript(
              userScript: UserScript(source: """
                async function wait() {
                  while(!document.getElementsByClassName('ScLoadingSpinnerCircle-sc-qd487d-1')[0]) {
                    await new Promise(r => setTimeout(r, 500));
                  }
                  while(document.getElementsByClassName('ScLoadingSpinnerCircle-sc-qd487d-1')[0]) {
                    await new Promise(r => setTimeout(r, 500));
                  }
                  ${widget.autoplay ? """""" : """document.querySelector('[data-a-target="player-play-pause-button"]').click();"""}
                }
                wait();
              """, injectionTime: UserScriptInjectionTime.AT_DOCUMENT_END),
            ),
            widget.controller.updateState(TwitchState.loading),
          },
          onLoadStop: (_, __) => {
            widget.controller.updateState(TwitchState.done),
          },
          onLoadError: (_, __, ___, ____) => widget.controller.updateState(TwitchState.error),
        ),
      ),
    );
  }
}
